streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/CopartitionedTopicsEnforcer.java

```diff
@@ -64,20 +66,70 @@ public void enforce(final Set<String> copartitionGroup,
                             }));
 
         final int numPartitionsToUseForRepartitionTopics;
+        final Collection<InternalTopicConfig> internalTopicConfigs = repartitionTopicConfigs.values();
+
         if (copartitionGroup.equals(repartitionTopicConfigs.keySet())) {
-            // If all topics for this co-partition group is repartition topics,
-            // then set the number of partitions to be the maximum of the number of partitions.
-            numPartitionsToUseForRepartitionTopics = getMaxPartitions(repartitionTopicConfigs);
+            if (internalTopicConfigs.stream().anyMatch(this::isMutableRepartitionTopicConfig)) {
+                // If all topics for this co-partition group is repartition topics,
+                // then set the number of partitions to be the maximum of the number of partitions.
+                numPartitionsToUseForRepartitionTopics = getMaxPartitions(repartitionTopicConfigs);
+            } else {
+                // if all repartition topics are immutable repartition topics,
+                // we must ensure that they all have same number of partitions.
+                numPartitionsToUseForRepartitionTopics = validateAndGetNumOfPartitionsOfImmutableTopics(repartitionTopicConfigs,
+                                                                                                        internalTopicConfigs);
+            }
         } else {
             // Otherwise, use the number of partitions from external topics (which must all be the same)
             numPartitionsToUseForRepartitionTopics = getSamePartitions(nonRepartitionTopicPartitions);
-
         }
 
         // coerce all the repartition topics to use the decided number of partitions.
-        for (final InternalTopicConfig config : repartitionTopicConfigs.values()) {
-            config.setNumberOfPartitions(numPartitionsToUseForRepartitionTopics);
+        for (final InternalTopicConfig config : internalTopicConfigs) {
+            if (isMutableRepartitionTopicConfig(config)) {
+                config.setNumberOfPartitions(numPartitionsToUseForRepartitionTopics);
+            } else {
+                final int numberOfPartitionsOfInternalTopic = config.numberOfPartitions().get();
+
+                if (numberOfPartitionsOfInternalTopic != numPartitionsToUseForRepartitionTopics) {
```

Comment:
To what extent is this check different from the check above when we call `validateAndGetNumOfPartitionsOfImmutableTopics` -- or can we remove it here as it's redundant?

Reply:
This case check covers the case when number of partitions do not match between immutable repartition topics (aka created via `repartition` operation) and _ordinary_ repartition topics. But considering your comment below, different logic is needed here.